### PR TITLE
Clean up declaration checker's modeling of generic arguments and parameter lists

### PR DESCRIPTION
**Pull request by slavapestov**
This is the preliminary cleanup for https://github.com/apple/swift/pull/33102.

Mostly NFC, but fixes <rdar://problem/58813746>, <https://bugs.swift.org/browse/SR-13004>.

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Review by CodaFi — APPROVED** [review]

**Review by AnthonyLatsis — APPROVED** [review]
Happy to see the listener go!

File: lib/Sema/TypeCheckGeneric.cpp [hunk 1]
```diff
@@ -631,9 +634,12 @@ GenericSignatureRequest::evaluate(Evaluator &evaluator,
                                              : diag::where_nongeneric_ctx);
       return nullptr;
     }
+  }
 
+  if (!genericParams && where)
     allowConcreteGenericParams = true;
-  } else {
+
+  if (!genericParams && !where) {
```
[the remark this answers is not in the document]
Comment (slavapestov) [reply lib/Sema/TypeCheckGeneric.cpp:642]: It’s a flow re-arrangement. It’s now a valid state to have both a GenericParamList *and* a TrailingWhereClause.

**Pull request merged**